libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -0,0 +1,3571 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE ImpredicativeTypes #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+module Constrained.Base where
```

Comment:
This module should probably be broken into multiple smaller modules at some point

Reply:
That's unfortunately very difficult to do. I've tried, but every time I keep coming back to having to put it back together because there is some circular dependency that you can't break nicely.

The core issue is that `Pred` depends on `Term` which depends on `HasSpec` which depends on `Generics` which depends `Spec` which depends on `Pred`.

However, I'm also willing to entertain the idea of doing something "dependency injectiony" to break this chain. But it would make the code base more complex.

Reply:
(that cycle I gave is not the only example, as you can imagine)

Reply:
Yeah, no need to do this now, maybe some time in the future.

Reply:
I'll put in a comment explaining the problem in brief.